src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -702,5 +703,38 @@ private static string GetPropertyName(MemberInfo property)
 
             return property.Name;
         }
+
+        private static IEnumerable<IConfigurationSection> GetChildrenOrSelf(IConfiguration config)
+        {
+            if (!ShouldBindSingleElementsToArray())
+            {
+                return config.GetChildren();
+            }
+
+
+            IEnumerable<IConfigurationSection> children;
+            // If configuration's children is an array, the configuration key will be a number
+            if (config.GetChildren().Any(a => long.TryParse(a.Key, out _)))
```

Comment:
@HaoK let us know if you know of a better way for this condition

Reply:
If any of the children is an integer (key:999), this assumes the actual object is an array. 

In real world, either all children are integers or all are not. But doing "Any" prevents usages in weird cases where someone used both things
having `key:a, key:0` doesn't make sense but i am assuming that it is an array (as in the old behavior)

Reply:
So the old behavior always used the children correct?  Why does this need to check if any children are a number?  What doesn't work if this is just returning the children if there are any, otherwise the config.

Reply:
This used to always return children for arrays/collections. But as part of my change, I might have to return the parent object or children object (when it contains array as a children vs non-array).
If children are a number, that means the children are an array and we select that. Otherwise, we select the parent object as a single-element array.
When there are nested objects (keys - "a:b", "a:c"), there are children but they cannot be bound. You have to select the parent object ("a" itself) for the single-element array to be bound.